packages/vite/src/node/plugins/css.ts

```diff
@@ -949,39 +881,131 @@ async function compileCSS(
+      const rawPostcssMap = postcssResult.map.toJSON()
+
+      const postcssMap = await formatPostcssSourceMap(
+        // version property of rawPostcssMap is declared as string
+        // but actually it is a number
+        rawPostcssMap as Omit<RawSourceMap, 'version'> as ExistingRawSourceMap,
+        cleanUrl(id)
+      )
+
+      map = combineSourcemapsIfExists(cleanUrl(id), postcssMap, preprocessorMap)
     }
-    throw e
+
+    css = postcssResult.css
   }
 
-  if (!devSourcemap) {
-    return {
-      ast: postcssResult,
-      code: postcssResult.css,
-      map: { mappings: '' },
-      modules,
-      deps
-    }
+  const lightningCss = await import('lightningcss')
+  let dependencies, exports
+  if (styleAttrRE.test(id)) {
+    const res = lightningCss.transformStyleAttribute({
+      filename: id,
+      code: Buffer.from(css),
+      analyzeDependencies: true,
+      minify: !!config.build.minify
+    })
+    dependencies = res.dependencies
+    code = res.code.toString()
+  } else {
+    const res = await lightningCss.bundleAsync({
+      filename: id,
+      resolver: {
+        read(filename) {
+          if (filename === id) {
+            // TODO: add source map inline comment
+            return css
+          }
+
+          try {
+            return fs.readFileSync(filename, 'utf8')
+          } catch (err) {
+            // TODO: handle external urls
+            return ''
+          }
+        },
+        async resolve(id, from) {
+          const publicFile = checkPublicFile(id, config)
+          if (publicFile) {
+            return publicFile
+          }
+
+          const resolved = await atImportResolvers.css(id, from)
+
+          if (resolved) {
+            return path.resolve(resolved)
+          }
+          return id
+        }
+      },
+      sourceMap: devSourcemap,
+      analyzeDependencies: true,
+      cssModules:
+        isModule && typeof modulesOptions?.generateScopedName === 'string'
+          ? {
+              pattern: modulesOptions?.generateScopedName, // TODO: function not supported.
+              dashedIdents: false
+            }
+          : isModule,
+      minify: !!config.build.minify
+      // TODO: set browser targets to enable transpilation
+    })
+
+    dependencies = res.dependencies
+    exports = res.exports
+    code = res.code.toString()
+    map = res.map ? JSON.parse(res.map.toString()) : undefined
   }
 
-  const rawPostcssMap = postcssResult.map.toJSON()
+  if (dependencies) {
+    for (const dep of dependencies) {
+      // TODO: do we need to add to deps?
 
-  const postcssMap = await formatPostcssSourceMap(
-    // version property of rawPostcssMap is declared as string
-    // but actually it is a number
-    rawPostcssMap as Omit<RawSourceMap, 'version'> as ExistingRawSourceMap,
-    cleanUrl(id)
-  )
+      if (dep.type === 'url') {
+        const resolved = await urlReplacer(dep.url, dep.loc.filePath)
+        code = code.replaceAll(dep.placeholder, resolved)
```

Comment:
This could potentially be done via a new API in Lightning CSS instead of replacing with placeholders.